### ISSUE TITLE
Variant aware dependency resolution fails to select a secondary variant with the correct attribute

### ISSUE THREAD

**Issue by mkobit** [gradle/gradle#16368]
### Expected Behavior

Dependency selection should select the outgoing variant from the dependency in the other project, regardless of which configuration or configuration variant includes the attributes for it.

### Current Behavior

Dependency resolution fails and the variant is not selected.

### Context

Attempting to break a part a multi-project build that has multiple configurations with explicit configuration dependencies on each project `dep(project(":someproj", "theConfig"))`.

### Steps to Reproduce 

* Gradle `6.8.3`
* `settings.gradle.kts`
    ```kts
    include("producer")
    include("consumer")
    ```

* `build.gradle.kts`
    ```kts
    tasks {
        wrapper {
            gradleVersion = "6.8.3"
        }
    }
    ```

* `producer/build.gradle.kts`
    ```kts
    plugins {
        `lifecycle-base`
    }
    
    val myUsage = Attribute.of("myUsage", Usage::class.java)
    val myUsageValue = objects.named(Usage::class, "myAttributeName")
    
    val outputFile = layout.buildDirectory.file("generatedFile.txt")
    
    val makeMyFile by tasks.registering {
        outputs.file(outputFile)
        doFirst {
            outputFile.get().asFile.writeText("hello world")
        }
    }
    configurations {
        create("myConfig") {
            isCanBeConsumed = true
            isCanBeResolved = false
            outgoing {
                variants {
                    create("myVariants") {
                        artifact(outputFile) {
                            builtBy(makeMyFile)
                        }
                        attributes {
                            attribute(myUsage, myUsageValue)
                        }
                    }
                }
            }
        }
    }
    ```

* `consumer/build.gradle.kts`
    ```kts
    plugins {
        `lifecycle-base`
    }
    
    val myUsage = Attribute.of("myUsage", Usage::class.java)
    val myUsageValue = objects.named(Usage::class, "myAttributeName")
    
    val myConsumer by configurations.creating {
        attributes {
            attribute(myUsage, myUsageValue)
        }
    }
    
    dependencies {
        myConsumer(project(":producer"))
    }
    
    val showConsumerFiles by tasks.registering {
        inputs.files(myConsumer)
        doFirst {
            println("myConsumer: ")
            myConsumer.asFileTree.forEach { println(" $it") }
        }
    }
    ```

* Running `./gradlew "consumer:showConsumerFiles` results in error
    ```
    Could not resolve all task dependencies for configuration ':consumer:myConsumer'.
    > Could not resolve project :producer.
      Required by:
          project :consumer
      > No matching configuration of project :producer was found. The consumer was configured to find attribute 'myUsage' with value 'myAttributeName' but:
          - Configuration 'myConfig':
              - Other compatible attribute:
                  - Doesn't say anything about myUsage (required 'myAttributeName')
    ```

### Your Environment

Build scan URL: https://gradle.com/s/qqljj2uorulqc

**Comment by ljacomet** [comment]
This rings a bell.

Can you try registering the attribute in the attribute schema as documented in example 2 of [this section of the documentation](https://docs.gradle.org/6.8.3/userguide/variant_attributes.html#creating_attributes_in_a_build_script_or_plugin) ?

You will have to do that for the producer and the consumer.

**Comment by mkobit** [comment]
Same issue when adding `dependencies.attributesSchema.attribute(myUsage)` to both `consumer` and `producer` projects

scan: https://gradle.com/s/ycp5v7qyj32gk

**Comment by vlsi** [comment]
1. The outgoing variant seems to be declared just fine

```
% ./gradlew :producer:OutgoingVariants

> Task :producer:outgoingVariants
--------------------------------------------------
Variant myConfig
--------------------------------------------------
Secondary variants (*)
    - Variant : myVariants
       - Attributes
          - myUsage = myAttributeName
       - Artifacts
          - build/generatedFile.txt (artifactType = txt)


(*) Secondary variants are variants created via the Configuration#getOutgoing(): ConfigurationPublications API which also participate in selection, in addition to the configuration itself.
```

---

However, it looks like the problem is resolved if the outgoing configuration declares at least one attribute.

For instance, the following works for the consumer:

```kotlin
configurations {
    create("myConfig") {
        isCanBeConsumed = true
        isCanBeResolved = false
        attributes {
            attribute(myUsage, myUsageValue) // <-- attributes at the configuration level solves the case
        }
        outgoing {
            variants {
                create("myVariants") {
                    artifact(outputFile) {
                        builtBy(makeMyFile)
                    }
                }
            }
        }
    }
}
```

=>

```
% ./gradlew :producer:OutgoingVariants :consumer:showConsumerFiles

> Task :producer:outgoingVariants
--------------------------------------------------
Variant myConfig
--------------------------------------------------
Capabilities
    - gr16368:producer:unspecified (default capability)
Attributes
    - myUsage = myAttributeName

Secondary variants (*)
    - Variant : myVariants
       - Attributes
          - myUsage = myAttributeName
       - Artifacts
          - build/generatedFile.txt (artifactType = txt)


(*) Secondary variants are variants created via the Configuration#getOutgoing(): ConfigurationPublications API which also participate in selection, in addition to the configuration itself.

> Task :consumer:showConsumerFiles
myConsumer:
.../gr16368/producer/build/generatedFile.txt
```

---

`org.gradle.usage = native-runtime` attribute at the configuration level resolves the case as well (it is not used here, but it heals the resolution somehow):

```
configurations {
    create("myConfig") {
        isCanBeConsumed = true
        isCanBeResolved = false
        attributes {
            attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.NATIVE_RUNTIME))
        }
        outgoing {
            variants {
                create("myVariants") {
                    artifact(outputFile) {
                        builtBy(makeMyFile)
                    }
                    attributes {
                        attribute(myUsage, myUsageValue)
                    }
                }
            }
        }
    }
}
```


```
% ./gradlew :producer:OutgoingVariants :consumer:showConsumerFiles

> Task :producer:outgoingVariants
--------------------------------------------------
Variant myConfig
--------------------------------------------------
Capabilities
    - gr16368:producer:unspecified (default capability)
Attributes
    - org.gradle.usage = native-runtime

Secondary variants (*)
    - Variant : myVariants
       - Attributes
          - myUsage          = myAttributeName
          - org.gradle.usage = native-runtime
       - Artifacts
          - build/generatedFile.txt (artifactType = txt)


(*) Secondary variants are variants created via the Configuration#getOutgoing(): ConfigurationPublications API which also participate in selection, in addition to the configuration itself.

> Task :consumer:showConsumerFiles
myConsumer:
.../gr16368/producer/build/generatedFile.txt
```

---

"dummy" value at the configuration level confuses the resolver:

```kotlin
configurations {
    create("myConfig") {
        isCanBeConsumed = true
        isCanBeResolved = false
        attributes {
            attribute(myUsage, objects.named(Usage::class, "dummy")) // <-- look here, dummy value
        }
        outgoing {
            variants {
                create("myVariantsw2") {
                    artifact(outputFile) {
                        builtBy(makeMyFile)
                    }
                    attributes {
                        attribute(myUsage, myUsageValue)
                    }
                }
            }
        }
    }
}
```

=>

```
% ./gradlew :producer:OutgoingVariants :consumer:showConsumerFiles

> Task :producer:outgoingVariants
--------------------------------------------------
Variant myConfig
--------------------------------------------------
Capabilities
    - gr16368:producer:unspecified (default capability)
Attributes
    - myUsage = dummy

Secondary variants (*)
    - Variant : myVariantsw2
       - Attributes
          - myUsage = myAttributeName
       - Artifacts
          - build/generatedFile.txt (artifactType = txt)


(*) Secondary variants are variants created via the Configuration#getOutgoing(): ConfigurationPublications API which also participate in selection, in addition to the configuration itself.

...

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':consumer:showConsumerFiles'.
> Could not resolve all task dependencies for configuration ':consumer:myConsumer'.
   > Could not resolve project :producer.
     Required by:
         project :consumer
      > No matching variant of project :producer was found. The consumer was configured to find attribute 'myUsage' with value 'myAttributeName' but:
          - Variant 'myConfig' capability gr16368:producer:unspecified:
              - Incompatible because this component declares attribute 'myUsage' with value 'dummy' and the consumer needed attribute 'myUsage' with value 'myAttributeName'
```

**Comment by vlsi** [comment]
TL;DR: it looks like there are two issues:
1) The implementation does not support the case when the configuration itself declares no attributes
2) The implementation does not handle the case when a variant-specific attribute value is different from the configuration value

**Comment by vlsi** [comment]
The behavior is the same for 6.8.1 and 7.0-milestone-1
